### DISCUSSION TITLE
Infinite Loading Wordpress

### DISCUSSION THREAD

**Discussion by lorenzo-moni** [swup/swup#745] [Help] [answered]
Hi, I'm having an issue implementing infinite loading function in a WordPress swup website.
the problem is that I'm trying to change the URL of the page in relation to the posts page that I am dynamically loading with js. When I change page URL with history.replaceState( null, null, $(this).attr("data-page") ) and I click on a post everything is right, the issue comes out when I need to come back to the post page using browser history. 

I think that the problem is related to the cache and the fact that I'm not updating it.
How can I do to update the cache of the page to let it adapt to the new content and URL?

This is my swup config.
```js
const options = {
    animateHistoryBrowsing: true,
    animationSelector: '[class*="transition-"]',
    containers: ["#swup"],
    cache: true,
    link_selector:'a[href^="/"]:not([data-no-swup]), a[href^="' + window.location.origin + '"]:not([data-no-swup])',
    skipPopStateHandling: function(event) {
      if (event.state && event.state.source == "swup") {
        return false;
      }
      return true;
    },
    plugins: [new SwupBodyClassPlugin()]
};
```
**Before creating this issue, did you think of...**:
- [x] Have you checked closed issues for similar/related problems.
- [x] Have you provided all helpful information available? 
- [ ] Have you considered creating a demo so we can help you better?

**Comment by AugustMiller** [comment]
Swup only keeps track of the pages it knows about. If you modify the history stack from outside Swup, things can get unpredictable.

Is it not possible to use the `loadPage()` method to swap your content in?

From the docs:

```js
swup.loadPage({
  url: $(this).attr("data-page"), // Substituted with your source—this would be inside an event handler!
  method: 'GET', // method of request (defaults to "GET")
  data: data, // data passed into XMLHttpRequest send method
  customTransition: '' // name of your transition used for adding custom class to html element and choosing custom animation in swupjs (as setting data-swup-transition attribute on link)
});
```

**Comment by AugustMiller** [comment]
Ah, disregard. I thought you were loading the individual _posts_ manually, but you're loading more content into the feed/index.

There may be a way to dynamically push entries into Swup's cache/history, but it's a ways off the beaten path, and not part of the public API.

Does your infinite scroll UI have "paginated" URLs, or is there any way to distinguish how far through the list of posts the user is (supposing someone is just using their browser's forward/back buttons, without Swup or AJAX)?

**Comment by lorenzo-moni** [comment]
Yes, the Infinite scroll has paginated URLs and you can clearly distinguish which section of the page the user was before clicking on the post and leaving.
The page URL is `http://domain/news/` and gets updated when the user scrolls on the page. The script has an isVisible function that changes the URL with

`history.replaceState( null, null, $(this).attr("data-page") );
return(false);`

When a block of news is visible the URL change in `http://asd-fornacette.local/news/page/2/`, `http://asd-fornacette.local/news/page/3/` etc.

I'm trying to update this new URL even into Swup cache so I can come back to that portion of the page.

If I disable Swup the script goes perfectly but otherwise I'm not able to figure out a solution.
Thanks in advance for helping.

**Comment by gmrchk** [comment]
Hi guys, interesting problem. If I understand correctly: 

* Page is loaded and swup works as expected.
* The page URL is updated (with replace) as you scroll, but swup doesn't know about the content updates.
* When you open a post and go back, swup doesn't know the URL, so it loads whatever is on the last URL updated by IS library, which is only part of the page we were viewing before.

Correct? If that's the case, what we could do is do the following every time the IS library has loaded the next page:
1. Remove the current page from the cache. Calling `swup.cache.remove('/news')` should do that. [Here](https://swup.js.org/api/cache)'s a cache docs.
2. Add the page content to swup cache manually with this updated content by IS. Swup does some manually adding as well over [here](https://github.com/swup/swup/blob/master/src/index.js#L120) on the first load, so you can get inspired there.

**Comment by hirasso** [comment]
Hi there! I see that I'm late to the party, but anyways:

I have implemented something like this in quite many websites I have built. These websites didn't use swup, yet (instead I was using my own home-grown page transition framework), but I am planning to switch to swup. Here's *roughly* how this works in my setup:

1. The wrapper of the infinitely scrolling element needs an `id` attribute, let's say `#infinite-scroll`
2. Loading of subsequent pages from infinite-scroll doesn't change the URL
3. Every time I load more items (let's say their selector is `.infinite-scroll__item`), I insert the new items from the server response into the current cache entry. 

In the case of `swup`, that would mean something like:

1. Parse the `swup.cache.getCurrentPage().originalContent` 
2. Insert the new infinite-scroll items after the last item inside the infinite-scroll element (the selector could be something example `#infinite-scroll .infinite-scroll__item:last`)
3. Update both the `originalContent` as well as the `blocks` entry in the swup cache entry accordingly to the new content

The great thing about this would be that the [scroll plugin](https://swup.js.org/plugins/scroll-plugin) would be able to actually scroll back to the correct scroll position when browser-navigating back to the infinite-scroll page, since all loaded items are persisted in the cache. I will have to build this functionality in the next project, should I choose to use swup for it. If I do, I will report back with my findings.

@gmrchk , could you point me to the internal swup functions you use to parse the DOM and to extract the blocks from it? I would love to be able to re-use as much swup intelligence as possible for this 🤓💯

**Comment by drivebass** [comment]
> Hi guys, interesting problem. If I understand correctly:
> 
>     * Page is loaded and swup works as expected.
> 
>     * The page URL is updated (with replace) as you scroll, but swup doesn't know about the content updates.
> 
>     * When you open a post and go back, swup doesn't know the URL, so it loads whatever is on the last URL updated by IS library, which is only part of the page we were viewing before.
> 
> 
> Correct? If that's the case, what we could do is do the following every time the IS library has loaded the next page:
> 
>     1. Remove the current page from the cache. Calling `swup.cache.remove('/news')` should do that. [Here](https://swup.js.org/api/cache)'s a cache docs.
> 
>     2. Add the page content to swup cache manually with this updated content by IS. Swup does some manually adding as well over [here](https://github.com/swup/swup/blob/master/src/index.js#L120) on the first load, so you can get inspired there.

Hi @gmrchk, possible to give details on how to add the page content to swup cache manually?

**Comment by hirasso** [comment, marked as answer]
With the [recent release of swup 4](https://swup.js.org/announcements/swup-4/), manipulating the cache became much simpler. Suppose you have the following HTML that will render a paginated list of blog posts (using [Astro](https://astro.build/) here):

```html
<ul class="posts">
  {
    page.data.map((post: any) => (
      <li class="post">
        <a href={`/blog/${post.slug}/`}>{post.data.title}</a>
      </li>
    ))
  }
</ul>
<ul class="pagination">
  {page.url.prev ? <li><a class="pagination__prev" href={page.url.prev}>Previous</a></li> : null}
  {page.url.next ? <li><a class="pagination__next" href={page.url.next}>Next</a></li> : null}
</ul>
```
This is how you could keep swup's cache in sync with your infinite scroll (using [infinite-scroll](https://infinite-scroll.com/)):
```typescript
import Swup from "swup";
import SwupPreloadPlugin from "@swup/preload-plugin";
import InfiniteScroll from "infinite-scroll";

const nextLinkSelector = '.pagination__next';

const swup = new Swup({
  plugins: [new SwupPreloadPlugin()] // will put the first page into the cache automatically
});

/**
 * Injects new items loaded by infinite-scroll into swup's cache entry of the current page
 */
function updateCache(containerSelector: string, items: NodeList, nextLink: Element) {
  const url = swup.getCurrentUrl();
  const cachedPage = swup.cache.get(url);
  if (!cachedPage) return;

  const cachedDocument = new DOMParser().parseFromString(cachedPage.html, 'text/html')
  const container = cachedDocument.querySelector(containerSelector);
  if (!container) return;

  // Update the items
  const clonedItems = [...items].map(item => item.cloneNode(true));
  container.append(...clonedItems);

  // Update the next link
  if (nextLink) {
    cachedDocument.querySelector(nextLinkSelector)?.replaceWith(nextLink.cloneNode(true));
  }

  // Save the modified html as a string in the cache entry
  cachedPage.html = cachedDocument.documentElement.outerHTML;
  swup.cache.update(url, cachedPage);
}

/**
 * Initialize infinite scroll (executed on selected swup hooks further down)
 */
function initInfiniteScroll() {
  const el = document.querySelector(".posts");
  if (!el) return;

  const nextLink = document.querySelector(nextLinkSelector);
  if (!nextLink) return;

  const infScroll = new InfiniteScroll(el, {
    path: nextLinkSelector,
    append: ".post",
    history: false,
    prefill: true,
  });

  infScroll.on("append", (doc: Document, _path: string, items: NodeList) => {
    // Get the next link. If there is no more available, replace it with an empty <span>
    const nextLink = doc.querySelector(nextLinkSelector) ?? doc.createElement('span');
    document.querySelector(nextLinkSelector)?.replaceWith(nextLink);
    // Update the cache
    updateCache(".posts", items, nextLink);
  });
}

/**
 * When we visit the page with the infinite scroll directly, that page's cache isn't there immediately. 
 * Preload Plugin takes care of (re-)preloading a fresh copy from the server and putting the result in swup's cache.
 * By hooking *once* into "page:preload", we can initialize our infinite scroll after the cached version of the page has arrived, so 
 * that we can safely append new items as they arrive.
 */
swup.hooks.once("page:preload", initInfiniteScroll)
/** Also init infinite scroll on every page view */
swup.hooks.on("page:view", initInfiniteScroll);


```

Please note that this example uses infinite scroll in a way that **doesn't update the browser url**. If you would like to support updating the browser URL as well, you could either 

- use `swup.options.resolveUrl` to cache only one page for all infinite scroll urls
- update the cache for each infinite scroll URL separately
